### PR TITLE
optimize isEmptyDeep

### PR DESCRIPTION
**Pull request by drzhbe**
**TL;DR** `isEmptyDeep` is now 300 times faster

Function `isEmptyDeep` suffered from 2 problems:
1. Unnecessary memory allocations
2. No early return

And since it's used in the snapping code, whenever you have many or just big geometries, the geometry editing is janky.

In this PR I optimized the `isEmptyDeep` function and added the tests for it.

Video explanation:
https://www.loom.com/share/e70379b76e5141d984e773b70dd47921

Benchmark:
https://jsbench.me/h8l8dv6d7h/2

### CODE REVIEW

File: cypress/integration/helpers.spec.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    expect(isEmptyDeep([[[undefined], [null], '']])).to.be.true;
+  });
+
+  it('should return false for a non-empty nested arrays', () => {
+    expect(isEmptyDeep([[[1]]])).to.be.false;
+  });
+
+  it('should return false if there is at least one non-empty value', () => {
+    expect(isEmptyDeep([[[null], [undefined, 1]]])).to.be.false;
+  });
```
Comment (Falke-Design) [inline cypress/integration/helpers.spec.js:22]: Can you please add tests with no nested arrays `isEmptyDeep([null, undefined, 1])` and `isEmptyDeep([null, undefined, ''])`

File: src/js/helpers/index.js [hunk 2]
```diff
@@ -13,13 +13,22 @@ export function getTranslation(path) {
...
+  const hasValues = (list) => {
+    for (let i = 0; i < list.length; i += 1) {
+      const item = list[i];
 
-  return !flatten(l).length;
+      if (Array.isArray(item)) {
+        if (hasValues(item)) {
+          return true;
+        }
+      } else if (item != null && item !== '') {
```
Comment (Falke-Design) [inline src/js/helpers/index.js:24]: Where is `undefined` checked? I see that the tests are working but I don't get where `undefined` is handled 😄

**Review by Falke-Design — CHANGES_REQUESTED** [review]
Wow that is great! Thank you!

Please add a few more tests like requested and then I think we can merge this

File: src/js/helpers/index.js (hunk 2 above)
Comment (drzhbe) [reply src/js/helpers/index.js:24]: `a == null` is the same as `a === undefined || a === null`. I can make it explicit

Comment (Falke-Design) [reply src/js/helpers/index.js:24]: ah yeah sure, I didn't noticed that it only has one `=`. Yes please make it explicit, thanks

**Comment by drzhbe** [comment]
@Falke-Design I've added the requested tests and made `=== undefined && === null` explicit
The tests look pretty tidy to me, wdyt?
Screen Shot 2022-09-23 at 9 35 18 pm

**Review by Falke-Design — APPROVED** [review]
This is really a nice improvement! Thank you very much!

@codeofsumit will make merge before the next release

**Comment by drzhbe** [comment]
@Falke-Design Thank you for the review! Do you know when approximately next release is gonna happen? Is it weeks/months away?

**Comment by Falke-Design** [comment]
Maybe with the end of oktober. I need first to fix most of the bugs but currently I have not much time

**Comment by drzhbe** [comment]
@Falke-Design I simplified the function a bit, removed the inner function `hasValues`. Now `isEmptyDeep` itself is recursive.

**Comment by Falke-Design** [comment]
I'm thinking about, if we should rename the function to `hasValues`, it would make it easier to understand the code in the function. Or does it not matter? What do you @drzhbe think?

**Comment by drzhbe** [comment]
@Falke-Design I agree the function reads a bit more straightforward now. And the name is shorter. Good suggestion 👍

**Review by codeofsumit — APPROVED** [review]
Amazing !!
